### ISSUE TITLE
Is there any method to unload a model from memory?

### ISSUE THREAD

**Issue by MoyanZitto** [keras-team/keras#11015]
It turns out once we load a model to memory, there's no method to unload it except to kill the process.

```python
model = load_model(model_path, compile=False)
x = input()
```
we use input() to block the process, so that we can see the memory usage through "top" or some other tools. Now, the model was loaded to the memory.

How would you think to release the memory?

Here're some possible solutions:

(1) set model=None so that GC will handle it.
(2) del model, it's like (1)
(3) use K.clear_session() or tf.reset_defualt_graph() or something like that.
(4) use (1) or (2) or (3) , or any combination，and call gc.collect() manually.

But none of above works. sys.getrefcount() would returns a very large (10k+) number after we set model=None.

I did't expect that this task would be so hard.

BTW, I noticed there're many unreachable objects just after we load the model, is that ok?

In my machine:

```python
model = load_model(model_path, compile=False)
unreachable_obj = gc.collect()
print(unreachable_obj)
```

would result in 7774.

@fchollet Pls see this issue.

**Comment by jlherren** [comment]
Short answer: This is normal and expected and you shouldn't worry about it.

Longer answer: Python memory management is not so trivial and you can find many good articles about it by searching for "python memory management". Put simply, when python garbage collects objects because nothing references them anymore, it will not release that memory back to the OS. One reason is that it's often not even possible to do that. Another reason is that it's more efficient, because python will need to allocate memory for new objects again sooner or later anyway, so why not just keep it and reuse it.

Bottom line is usually this: Do not worry about it and don't try too hard to fight it. Unsetting a variable when you're done with it is often a good idea though; it's actually good practice to only use local variables for such large objects, not global ones. If you then still happen to run out of memory or if you suspect that your code really should use less memory, then please provide a full example that demonstrates the problem. In any case, this is likely less an issue with Keras than it is about python in general.

**Comment by jlherren** [comment]
Sometimes memory can and will be released back to the OS, but most often it won't. This is not specific to python, this is true for all processes. This is why long-running services (for example apache) employ a forking model where new processes are created and later killed over and over again.  If it matters so much to you to release memory as soon as possible, perhaps you can do something similar. 

It's still possible that there's a bug in keras, but for you example code above it is really quite normal for memory to not be released to the OS.

**Comment by farizrahman4u** [comment]
@MoyanZitto  Do all the models have the same config? i.e,  are they different versions of the same model with different weights?

**Comment by MoyanZitto** [comment]
@farizrahman4u No, I don't how other users will use the machine. Perhaps they're not using deep learning at all. 

Based on discussion at #2102 ,  if someone else can solve the problem by degrading keras, perhaps we should upgrade K.clear_session() so that it could work properly with latest tensorflow.

**Comment by farizrahman4u** [comment]
What is the latest version of TF and Keras that works? This is probably a bug in Keras topology (maintaining references to weights) rather than in `K.clear_session`. Lets look at the diff and fix it.

**Comment by berlinguyinca** [comment]
are there any updates on this?

**Comment by saintthor** [comment]
I am facing this problem now. In tf2.0, the keras model takes more memory than tf 1.14. 

I have to separate the process to two. one is for the model only. terminate it when memory exhausts.

**Comment by spearsem** [comment]
I am also facing this issue. `clear_session` is not a viable option for me as I have to load multiple Keras models at the same time during parallel executing unit tests.

**Comment by farizrahman4u** [comment]
I will look into it. Got some code to test?

**Comment by saintthor** [comment]
> I will look into it. Got some code to test?

https://github.com/tensorflow/tensorflow/issues/33412

here is the test code and model.
